[JSC] Optimize 8bit string array search by `Array#indexOf` in DFG/FTL

sosukesuzuki opened the pull request:
#### b68f6434c4f5badc76a61b2c1a778a86855a87e3
<pre>
[JSC] Optimize 8bit string array search by `Array#indexOf` in DFG/FTL
<a href="https://bugs.webkit.org/show_bug.cgi?id=286244">https://bugs.webkit.org/show_bug.cgi?id=286244</a>

Reviewed by Yusuke Suzuki.

The DFG and FTL for `Array#indexOf` calls `operationArrayIndexOfString`
to search string  from array of strings. However, for arrays of 8-bit
strings, this can be more optimized.

                                           TipOfTree                  Patched

array-prototype-indexOf-string          80.1512+-1.4965     ^     49.4774+-1.6997        ^ definitely 1.6200x faster
array-prototype-indexOf-string-16       84.6494+-0.0597           83.0947+-2.7614          might be 1.0187x faster

* JSTests/microbenchmarks/array-prototype-indexOf-string-16.js: Added.
* JSTests/microbenchmarks/array-prototype-indexOf-string.js: Added.
* JSTests/stress/array-prototype-indexOf-string-16.js: Added.
(shouldBe):
(test1):
(test2):
* JSTests/stress/array-prototype-indexOf-string-8.js: Added.
(shouldBe):
(test1):
(test2):
* JSTests/stress/array-prototype-indexOf-string-obj.js: Added.
(shouldBe):
(test1):
(test2):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileArrayIndexOf):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):

Canonical link: <a href="https://commits.webkit.org/289780@main">https://commits.webkit.org/289780@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8849cfaf6ff288735c0eb0ad0d07383ba4c35e18

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88030 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7546 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42428 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92978 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38780 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7927 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15724 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67949 "") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25688 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91032 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6073 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79645 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48319 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5848 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34055 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37887 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/80828 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76241 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34937 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94826 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/86806 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15198 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76805 "") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15453 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75501 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76044 "") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18690 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20439 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/18848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8208 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15216 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/109299 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14958 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26279 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18403 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16740 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (commented):
Can you add FTL too? Whenever something gets enabled, it must be optimized most in FTL.

Constellation reviewed (approved):
r=me

sosukesuzuki committed:
[JSC] Optimize 8bit string array search by `Array#indexOf` in DFG/FTL
https://bugs.webkit.org/show_bug.cgi?id=286244

Reviewed by Yusuke Suzuki.

The DFG and FTL for `Array#indexOf` calls `operationArrayIndexOfString`
to search string  from array of strings. However, for arrays of 8-bit
strings, this can be more optimized.

                                           TipOfTree                  Patched

array-prototype-indexOf-string          80.1512+-1.4965     ^     49.4774+-1.6997        ^ definitely 1.6200x faster
array-prototype-indexOf-string-16       84.6494+-0.0597           83.0947+-2.7614          might be 1.0187x faster

* JSTests/microbenchmarks/array-prototype-indexOf-string-16.js: Added.
* JSTests/microbenchmarks/array-prototype-indexOf-string.js: Added.
* JSTests/stress/array-prototype-indexOf-string-16.js: Added.
(shouldBe):
(test1):
(test2):
* JSTests/stress/array-prototype-indexOf-string-8.js: Added.
(shouldBe):
(test1):
(test2):
* JSTests/stress/array-prototype-indexOf-string-obj.js: Added.
(shouldBe):
(test1):
(test2):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileArrayIndexOf):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):

Canonical link: https://commits.webkit.org/289780@main
